### PR TITLE
Constraint Linear Operator

### PR DESCRIPTION
**Pull request by ESeNonFossiIo**

### CODE REVIEW

File: include/deal.II/lac/linear_operator.h [hunk 1]
```diff
@@ -1111,6 +1111,112 @@ linear_operator(const OperatorExemplar &operator_exemplar, const Matrix &matrix)
...
+          typename Domain = Range,
+          typename Matrix>
+LinearOperator<Range, Domain>
+constraints_linear_operator(const ConstraintMatrix &, const Matrix &);
+
+/**
+ * @relates LinearOperator
+ *
+ * constraints_linear_operator takes a constraint matrix @p cm and an exemplar
+ * matrix @p (this matrix could be the system matrix).
```
Comment (bangerth) [inline include/deal.II/lac/linear_operator.h]: "@p (..." -> "@p m (..."

File: include/deal.II/lac/linear_operator.h [hunk 2]
```diff
@@ -1111,6 +1111,112 @@ linear_operator(const OperatorExemplar &operator_exemplar, const Matrix &matrix)
...
+ * The result of this function is the LinearOperator associated to the constraint
+ * matrix.
```
Comment (bangerth) [inline include/deal.II/lac/linear_operator.h]: I don't think I actually know what this linear operator is. Can you be explicit about what it represents, given a set of constraints $x_i = \sum_j C_{ij} x_j + b_i$?

File: include/deal.II/lac/linear_operator.h [hunk 3]
```diff
@@ -1111,6 +1111,112 @@ linear_operator(const OperatorExemplar &operator_exemplar, const Matrix &matrix)
...
+LinearOperator<Range, Domain>
+constrained_linear_operator(const ConstraintMatrix &, const Matrix &);
+
+/**
+ * @relates LinearOperator
+ *
+ * Let @p m be a system matrix and assume we are interested in this problem
+ * constrained with a constraint matrix @p cm.
+ * This function returns the LinearOperator that replace the system matrix and
+ * is associated to this problem.
```
Comment (bangerth) [inline include/deal.II/lac/linear_operator.h]: same here. I think it would be useful to explicitly spell out what operation this LinearOperator does.

File: include/deal.II/lac/packaged_operation.h [hunk 4]
```diff
@@ -827,6 +827,54 @@ operator*(const PackagedOperation<Range> &comp,
...
+PackagedOperation<Range>
+constrained_rhs(const ConstraintMatrix &, const Matrix &, const Range &);
+
+/**
+ * @relates LinearOperator
+ *
+ * Let @p m be a system matrix, @p rhs a vector, and @p cm a constraint matrix.
+ *
+ * constrained_rhs returns the rhs associated to @p rhs of the constrained system
+ * induced by @p m and @p cm.
```
Comment (bangerth) [inline include/deal.II/lac/packaged_operation.h]: and again here

**Comment by drwells** [comment]
It looks like you wrote 'deatails' when you meant to write 'details' in a few (three) places.

File: include/deal.II/lac/linear_operator.h [hunk 5]
```diff
@@ -1111,6 +1111,130 @@ linear_operator(const OperatorExemplar &operator_exemplar, const Matrix &matrix)
...
+ * @relates LinearOperator
+ *
+ * constraints_linear_operator takes a constraint matrix @p cm and an exemplar
+ * matrix @p m (this matrix could be the system matrix).
+ * The result of this function is the LinearOperator associated to the constraint
+ * matrix.
+ * In detail, a constrained problem can be expressed in this form:
+ * Ct * S * C * x = Ct * ( rhs + S * k)
+ * Where:
+ *  - C has the role of constraint matrix
```
Comment (bangerth) [inline include/deal.II/lac/linear_operator.h]: I think you should still explain how this C relates to the constraints and inhomogeneities stored by the ConstraintMatrix class.

Specifically, the ConstraintMatrix class thinks of itself as a matrix of size $m x n$ where m=number of constraints, and n=number of dofs. On the other hand, you tend to think of C as an $n x n$ matrix.

File: include/deal.II/lac/linear_operator.h [hunk 6]
```diff
@@ -1111,6 +1111,130 @@ linear_operator(const OperatorExemplar &operator_exemplar, const Matrix &matrix)
...
+ *  - k is the vector representing constraints inhomogeneities.
+ *  [ a generic contraint can be expressed as x = Cx + k ]
```
Comment (bangerth) [inline include/deal.II/lac/linear_operator.h]: Just use round parentheses, and no spaces inside the parentheses.

File: include/deal.II/lac/linear_operator.h [hunk 7]
```diff
@@ -1111,6 +1111,130 @@ linear_operator(const OperatorExemplar &operator_exemplar, const Matrix &matrix)
...
+ *  - S is the system matrix
+ *  - rhs is the original right-hand-side
+ * This function returns a LinearOperator representing the matrix C.
+ *
+ * [ see M. S. Shephard:. Linear multipoint constraints applied via
+ * transformation as part of a direct stiffness assembly process., 1985.
+ * for more details ]
```
Comment (bangerth) [inline include/deal.II/lac/linear_operator.h]: Your periods '.' are in the wrong place. There shouldn't be one after the author + colon. There shouldn't be one after the title. There should be a comma after 1985. And there should be one after "details".

Instead of putting this into parentheses, just use `@see` to mark it up for doxygen.

Comment (bangerth) [inline include/deal.II/lac/linear_operator.h]: same in the other places where you reference the paper.

File: include/deal.II/lac/linear_operator.h (hunk 3 above)
Comment (ESeNonFossiIo) [inline include/deal.II/lac/linear_operator.h]: I wrote some lines. Should I be more verbose?
Should I copy and past this part under other functions?

File: include/deal.II/lac/linear_operator.h [hunk 8]
```diff
@@ -1111,6 +1111,140 @@ linear_operator(const OperatorExemplar &operator_exemplar, const Matrix &matrix)
...
+ *    (a generic contraint can be expressed as x = Cx + k)
+ *  - S is the system matrix
+ *  - rhs is the original right-hand-side
+ * This function returns a LinearOperator representing the matrix Ct * S * C.
+ *
+ * @note Suppose we have n dof and m constraints. W.l.o.g. we can assume that it
+ * is possible to express n-m variables in terms of remainder variables
+ * (constrained variables).
+ * Therefore, x_i = C_{i,j} x_j + k_i for j = 1, ..., n-m and i = 1, ..., n
+ * Notice that Ct * S * C  is a problem in R^{n-m}, remaineder variables are
```
Comment (bangerth) [inline include/deal.II/lac/linear_operator.h]: typo: remaineder

File: include/deal.II/lac/linear_operator.h [hunk 9]
```diff
@@ -1111,6 +1111,140 @@ linear_operator(const OperatorExemplar &operator_exemplar, const Matrix &matrix)
...
+ * treated solving x = 0 in order to have a well-posed problem on R^{n}.
```
Comment (bangerth) [inline include/deal.II/lac/linear_operator.h]: You need to put the `R^{n}` between dollar signs for this to be rendered as a formula. In that case, you can also write it as `${\mathbb R}^n$` to make it even prettier.

File: include/deal.II/lac/linear_operator.h [hunk 10]
```diff
@@ -1111,6 +1111,140 @@ linear_operator(const OperatorExemplar &operator_exemplar, const Matrix &matrix)
...
+ *  - k is the vector representing constraints inhomogeneities.
+ *    (a generic contraint can be expressed as x = Cx + k)
+ *  - S is the system matrix
+ *  - rhs is the original right-hand-side
+ * This function returns a LinearOperator representing the matrix Ct * S * C.
+ *
+ * @note Suppose we have n dof and m constraints. W.l.o.g. we can assume that it
+ * is possible to express n-m variables in terms of remainder variables
+ * (constrained variables).
+ * Therefore, x_i = C_{i,j} x_j + k_i for j = 1, ..., n-m and i = 1, ..., n
```
Comment (bangerth) [inline include/deal.II/lac/linear_operator.h]: I think that's backward. How about this wording:
```
To express the form of $C$, let us consider that it corresponds to a ConstraintMatrix object on $n$ degrees of freedom, of which $m\le n$ are constrained. Without loss of generality, let us assume that the first $m$ are the ones that are constrained, and have the form $x_i = \sum_{j=m+1}^n c_{ij} x_j + b_j$ for $i=1,\ldots, m$. Then the rest of the degrees of freedom could be written as $x_i=x_i$ for $i=m+1,\ldots, n$. The matrix $C$ is then that matrix that has entries $C_{ij}=c_{ij}$ for $1\le i \le m, m+1\le j\le n$, $c_{ii}=1$ for $m+1\le i \le n$, and $C_{ij}=0$ for all other elements.
```

File: include/deal.II/lac/linear_operator.h [hunk 11]
```diff
@@ -1111,6 +1111,154 @@ linear_operator(const OperatorExemplar &operator_exemplar, const Matrix &matrix)
...
+              {
+                types::global_dof_index pos = (*entries)[j].first;
+                v(pos) += u(i) * (*entries)[j].second;
+              }
+          }
+        else
+          v(i)=u(i);
+
+      }
+  };
```
Comment (tamiko) [inline include/deal.II/lac/linear_operator.h]: The ```vmult_add``` and ```Tvmult_add``` variants are missing. What about the following:
 * rename above variants to ```vmult_add``` and ```Tvmult_add``
 * replace ```v(i)=u(i)``` in line 1179 and 1199 by ```v(i)+=u(i)```
 * remove ```v(i)=0``` in line 1169 and 1189

After that you can implement ```vmult``` and ```Tvmult``` in two lines:
```
return_op.vmult = [vmult_add](Range &v, const Domain &u)
{
  v = 0.;
  vmult_add(v, u);
};

**Comment by tamiko** [comment]
@bangerth We would like to discuss the naming choices a bit. Currently we have
 * ```constraints_linear_operator``` that applies constraints stored in a ConstraintMatrix to a given vector.
 * ```constrained_linear_operator``` that uses above encapsulation to around a vmult with a matrix.
 * ```constrained_rhs``` that returns a PackagedOperation to compute the corresponding right hand side.

**Comment by bangerth** [comment]
I'm not particularly fond of the `constraints_linear_operator` name. This corresponds to the `ConstraintMatrix::distribute` operation, if I see this right (though the latter works in-place, whereas the linear operator requires a temporary vector).

All three of these require a `ConstraintMatrix` object -- why not make the function that create these things a member function of `ConstraintMatrix`? This would also make it easier to describe what they do in the documentation. For example, `constraints_linear_operator` could become `ConstraintMatrix::get_distribute_operation()` or `ConstraintMatrix::get_operator_for_distribute()` or similar, and in the documentation you would simply say something along the lines 
```
  This function returns a linear operator $O$ so that the operation $v=O*u$ (i.e., calling 
  <code>O.vmult (v,u)</code>) is equivalent to calling <code>v=u; constraint_matrix.distribute(v);</code>,
  where <code>constraint_matrix</code> is current object.
```

Similarly, the second one above may become `ConstraintMatrix::get_constrained_linear_operator(...)` and the last one `ConstraintMatrix::get_constrained_rhs_operation()` or similar.

**Comment by tamiko** [comment]
@bangerth I'm strongly against making this a member function. Those three functions are just thin wrappers around the public interface of ```ConstraintMatrix```. So there is no justification to enlarge the class signature at all.

**Comment by bangerth** [comment]
OK. Can you do something like `@relates ConstraintMatrix` to make sure there is a connection in the documentation?

**Comment by ESeNonFossiIo** [comment]
@tamiko "Those three functions are just thin wrappers around the public interface of ConstraintMatrix". They are not "just" thin wrappers... Think to the case you do not have an assembled system matrix (you know just how vmult works) or you do not what to assemble the whole constrained system matrix. In **ConstraintMatrix** you have to assemble the constraint matrix but not in this case.

**Comment by bangerth** [comment]
I'll leave this to you two to figure out. As a sidenote, if you created these operators inside `ConstraintMatrix`, you'd have access to the internal data structures, not just the public interface.

**Comment by tamiko** [comment]
@ESeNonFossiIo *ping*

Let's get this merged.

No, those functions do not belong into ConstraintMatrix - we do not need any internal data structures, further this leads to a namespace pollution in the constraint_matrix.h header.

**Comment by ESeNonFossiIo** [comment]
It is fine with me. Should I change something?

**Comment by tamiko** [comment]
@ESeNonFossiIo Do I have push access to your deal.II repository? Then, I'll rebase and possibly reword parts of the documentation a bit.

The only real thing missing is a good idea how to name the functions. Let me think about that.

**Comment by ESeNonFossiIo** [comment]
I have just granted you push access  to my deal.II repository.

**Comment by tamiko** [comment]
@bangerth I have restructured everything into its own header file ```constraint_linear_operator.h``` and I have renamed all functions to make its intentions clear.

@ESeNonFossiIo, @luca-heltai As you can see, I have restructured everything quite a bit. (The former state of this pull request was nowhere near functional, nor even compilable...). Now, the system that is solved is
```
(Ct * A * C + Id_c) x = C_t (b - A * k);
```
There is a problem with the testcase, though. It doesn't converge.

**Comment by ESeNonFossiIo** [comment]
@tamiko 
in *distribute_constraints_linear_operator* you should use
``` cpp
  return_op.vmult = [&constraint_matrix](Range &v, const Domain &u)
  {
    // v.locally_owned_elements() == complete_index_set (v.size())
    for(auto i : v.locally_owned_elements())
      {
        if (constraint_matrix.is_constrained(i))
          {
            v(i) = 0;
            const auto
            *entries = constraint_matrix.get_constraint_entries (i);
            for (types::global_dof_index j=0; j < entries->size(); ++j)
              {
                const auto pos = (*entries)[j].first;
                v(i) +=  u(pos) * (*entries)[j].second;
              }
          }
        else
          v(i) = u(i);
      }
  };

  return_op.Tvmult = [&constraint_matrix](Range &v, const Domain &u)
  {
    for(auto i : u.locally_owned_elements())
      {
        if (constraint_matrix.is_constrained(i))
          {
            v(i)=0;
            const auto
            *entries = constraint_matrix.get_constraint_entries (i);
            for (types::global_dof_index j=0; j < entries->size(); ++j)
              {
                types::global_dof_index pos = (*entries)[j].first;
                v(pos) += u(i) * (*entries)[j].second;
              }
          }
        else
          v(i)=u(i);

      }
  };
```
According to some tests on my laptop it seems to be the solution.

**Comment by ESeNonFossiIo** [comment]
By the way, you new implementation seems to be numerically better than the previous:
```
22: ! DEAL:cg::Starting value 0.1319444444
22: ! DEAL:cg::Convergence step 11 value 3.873045348e-09
22: ! DEAL:cg::Starting value 0.1319444444
22: ! DEAL:cg::Convergence step 11 value 3.873045348e-09
```
VS
```
22: ! DEAL:cg::Starting value 16.36993828
22: ! DEAL:cg::Convergence step 38 value 4.980526792e-07
22: ! DEAL:cg::Starting value 0.03428819444
22: ! DEAL:cg::Convergence step 231 value 9.972355928e-07
22: --- 1,8 ----
```

File: include/deal.II/lac/constraint_linear_operator.h [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+  return_op.Tvmult_add = [&constraint_matrix](Domain &v, const Range &u)
+  {
+    for (auto i : v.locally_owned_elements())
+      if (constraint_matrix.is_constrained(i))
+        v(i) += u(i);
+  };
+
+  // lambda capture expressions are a C++14 feature...
+  const auto vmult_add = return_op.vmult_add;
+  auto vmult = [vmult_add](Range &v, const Domain &u)
```
Comment (tamiko) [inline include/deal.II/lac/constraint_linear_operator.h]: @ESeNonFossiIo Much simpler. It must read ```return_op.vmult```, not ```auto vmult```... *sigh*

**Comment by tamiko** [comment]
@ESeNonFossiIo Please have a final look.

@bangerth I consider this ready to be merged now. Please have a look at the documentation.

**Comment by ESeNonFossiIo** [comment]
@tamiko 
It looks good! :D
I like very much your restyle of the code!

**Comment by tamiko** [comment]
@ESeNonFossiIo Nice to hear :-)
Looking forward for the distributed variant once this is merged  :-]

File: doc/doxygen/headers/constraints.h [hunk 13]
```diff
@@ -432,4 +432,59 @@
  * Either behavior can also be achieved by building two separate
  * ConstraintMatrix objects and calling ConstraintMatrix::merge function with
  * a particular second argument.
+ *
+ *
+ * <h3>Applying constraints indirectly with a LinearOperator</h3>
+ *
+ * Sometimes it is either not desirable, or not possible to directly
+ * condensate, or eliminate constraints from a system of linear equations.
```
Comment (bangerth) [inline doc/doxygen/headers/constraints.h]: condensate -> condense

File: doc/doxygen/headers/constraints.h [hunk 14]
```diff
@@ -432,4 +432,59 @@
...
+ * In particular if there is no underlying matrix object that could be
+ * condensed (or taken care of constraints during assembly). This is
+ * usually the case if the system is described by a LinearOperator.
+ *
+ * In this case we can solve the modified system
+ * @f[
+ *   (C^T A C + Id_c) \tilde x = C^T (b - A\,k)
+ * @f]
+ * instead [1]. Here, $A$ is a given (unconstrained) system matrix $A$ and
```
Comment (bangerth) [inline doc/doxygen/headers/constraints.h]: no need to repeat `$A$`

File: doc/doxygen/headers/constraints.h [hunk 15]
```diff
@@ -432,4 +432,59 @@
...
+ * $b$ the corresponding right hand side of a system of linear equations
+ * $A\,x=b$. The matrix $C$ describes the homogeneous part of the linear
+ * constraints stored in a ConstraintMatrix and the vector $k$ is the
+ * vector of corresponding inhomogeneities. More precisely, the
+ * ConstraintMatrix::distribute() operation applied on a vector $x$ is the
+ * operation
+ * @f[
+    x \leftarrow C\,x+k.
```
Comment (bangerth) [inline doc/doxygen/headers/constraints.h]: This assumes that $C$ can be thought of as a square matrix, but we've typically thought of a matrix that only has as many rows as there are constrained DoFs.

File: doc/doxygen/headers/constraints.h [hunk 16]
```diff
@@ -432,4 +432,59 @@
...
+ * @f]
+ * And finally, $Id_c$ denotes the identity on the subspace of constrained
+ * degrees of freedom.
+ *
+ * The corresponding solution of $A\,x=b$ that obeys boundary conditions is
```
Comment (bangerth) [inline doc/doxygen/headers/constraints.h]: boundary conditions -> these constraints

File: doc/doxygen/headers/constraints.h [hunk 17]
```diff
@@ -432,4 +432,59 @@
...
+ * // right_hand_side   - unconstrained and assembled right hand side
+ * // constraint_matrix - a ConstraintMatrix object
+ * // solver            - an appropriate, iterative solver
+ * // preconditioner    - a preconditioner
+ *
+ * const auto op_a = linear_operator(system_matrix);
+ * const auto op_amod = constrained_linear_operator(constraint_matrix, op_a);
+ * Vector<double> rhs_mod = constrained_right_hand_side(constraint_matrix,
+ *                                                      op_a,
+ *                                                      right_hand_side);
```
Comment (bangerth) [inline doc/doxygen/headers/constraints.h]: does `constrained_right_hand_side` result in a vector, or in a delayed `PackagedOperation`? In the latter case, if you don't assign it to a variable but put the expression right into the call of the solver, how often will the expression be evaluated?

File: doc/doxygen/headers/constraints.h [hunk 18]
```diff
@@ -432,4 +432,59 @@
...
+ *
+ * solver.solve(op_amod, solution, rhs_mod, preconditioner);
+ * constraint_matrix.distribute(solution);
+ * @endcode
+ *
+ * [1] M. S. Shephard: Linear multipoint constraints applied via
+ * transformation as part of a direct stiffness assembly process, 1985.
```
Comment (bangerth) [inline doc/doxygen/headers/constraints.h]: Just move the footnote into parentheses at the place where you use it.

File: doc/news/changes.h [hunk 19]
```diff
@@ -159,6 +159,14 @@ inconvenience this causes.
 <a name="general"></a>
 <h3>General</h3>
 <ol>
+  <li> New: constrained_linear_operator() and constrained_right_hand_side()
+  implemented that provide a generic mechanism of applying constraints to a
```
Comment (bangerth) [inline doc/news/changes.h]: remove "implemented that". We typically use present tense in changelog entries.

File: include/deal.II/lac/constraint_linear_operator.h [hunk 20]
```diff
@@ -0,0 +1,291 @@
...
+ *
+ * @author Mauro Bardelloni, Matthias Maier, 2015
+ *
+ * @relates LinearOperator
+ * @ingroup constraints
+ */
+template <typename Range, typename Domain>
+LinearOperator<Range, Domain>
+constrained_linear_operator(const ConstraintMatrix &constraint_matrix,
+                            const LinearOperator<Range, Domain> &linop)
```
Comment (bangerth) [inline include/deal.II/lac/constraint_linear_operator.h]: the function is called `constrained_linear_operator`, but the filename is `constraint_linear_operator.h`. Is this on purpose?

File: include/deal.II/lac/constraint_linear_operator.h [hunk 21]
```diff
@@ -0,0 +1,291 @@
...
+/**
+ * This function takes a ConstraintMatrix @p cm and an operator exemplar @p
+ * exemplar (this exemplar is usually a linear operator that describes the
+ * system matrix) and returns a LinearOperator object associated with the
+ * "homogeneous action" of the underlying ConstraintMatrix object:
+ *
+ * Applying the LinearOperator object on a vector <code>u</code> results in
+ * a vector <code>v</code> that stores the result of calling
+ * ConstraintMatrix::distribute() on <code>u</code> - with one important
+ * difference; inhomogeneities are note applied, but always treated as 0
```
Comment (bangerth) [inline include/deal.II/lac/constraint_linear_operator.h]: ; -> :

Comment (bangerth) [inline include/deal.II/lac/constraint_linear_operator.h]: I think it would actually be useful to state what the operator does for each non-constrained/constrained element of the vector.

File: doc/doxygen/headers/constraints.h (hunk 17 above)
Comment (tamiko) [inline doc/doxygen/headers/constraints.h]: > does constrained_right_hand_side result in a vector, or in a delayed PackagedOperation?

In a ```PackagedOperation```

> but put the expression right into the call of the solver, how often will the expression be evaluated?

Once.

Our solver classes are templated with the vector type they accept. The ```PackagedOperation``` will be implicitly converted to a vector upon invocation of ```solve``` and passed by value.

File: doc/doxygen/headers/constraints.h (hunk 15 above)
Comment (tamiko) [inline doc/doxygen/headers/constraints.h]: Is $x \leftarrow C\,x + k$ not clear enough? Shall I also mention that it is the linear part of the otherwise affine function ```distrbute```?

Otherwise I am a bit at a loss.

Comment (bangerth) [inline doc/doxygen/headers/constraints.h]: What I mean is that in https://www.dealii.org/developer/doxygen/deal.II/classConstraintMatrix.html we describe the matrix that is associated with a `ConstraintMatrix` object as having size `m` by `n` where `m` equals the number of constraints. So multiplying with this matrix would yield a vector of size `m`, but the way you explain the matrix `C` here is seems to me that it's `n` by `n` where the rows corresponding to unconstrained degrees of freedom have only the diagonal entry set to one. 

I'm just pointing out that we describe "the matrix associated with the ConstraintMatrix object" in different ways here and in the class documentation.

File: include/deal.II/lac/constrained_linear_operator.h [hunk 22]
```diff
@@ -0,0 +1,292 @@
...
+//@{
+
+
+/**
+ * This function takes a ConstraintMatrix @p constraint_matrix and an
+ * operator exemplar @p exemplar (this exemplar is usually a linear
+ * operator that describes the system matrix) and returns a LinearOperator
+ * object associated with the "homogeneous action" of the underlying
+ * ConstraintMatrix object:
+ *
```
Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: Did you mean to say something after the colon?

File: include/deal.II/lac/constrained_linear_operator.h [hunk 23]
```diff
@@ -0,0 +1,292 @@
...
+ * Applying the LinearOperator object on a vector <code>u</code> results in
+ * a vector <code>v</code> that stores the result of calling
+ * ConstraintMatrix::distribute() on <code>u</code> - with one important
+ * difference: inhomogeneities are note applied, but always treated as 0
```
Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: note -> not

File: include/deal.II/lac/constrained_linear_operator.h [hunk 24]
```diff
@@ -0,0 +1,292 @@
...
+  const LinearOperator<Range, Domain> &exemplar)
+{
+  LinearOperator<Range, Domain> return_op = exemplar;
+
+  return_op.vmult_add = [&constraint_matrix](Range &v, const Domain &u)
+  {
+    for (auto i : v.locally_owned_elements())
+      {
+        if (constraint_matrix.is_constrained(i))
+          {
```
Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: The algorithm here is not wrong, but it's almost certainly going to be pretty slow compared to the implementation in ConstraintMatrix::distribute itself. This may not matter all that much, but it's still worth asking: Is there a way to let that class do the computation, rather than duplicate the algorithm here?

File: include/deal.II/lac/constrained_linear_operator.h [hunk 25]
```diff
@@ -0,0 +1,292 @@
...
+          v(i)+=u(i);
+      }
+  };
+
+  // lambda capture expressions are a C++14 feature...
+  const auto vmult_add = return_op.vmult_add;
+  return_op.vmult = [vmult_add](Range &v, const Domain &u)
+  {
+    v = 0.;
+    vmult_add(v, u);
```
Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: This too is not wrong, but slow -- for every write operation you have to read v once. It shouldn't be too hard to specialize the algorithm above for a zero starting vector.

File: include/deal.II/lac/constrained_linear_operator.h [hunk 26]
```diff
@@ -0,0 +1,292 @@
...
+  };
+
+  return return_op;
+}
+
+
+/**
+ * Given a ConstraintMatrix @p constraint_matrix and an operator exemplar
+ * @p exemplar, return an LinearOperator that is the projection to the
+ * subspace of constrained degrees of freedom.
```
Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: Can you spell out what this means in terms of formulas?

File: include/deal.II/lac/constrained_linear_operator.h [hunk 27]
```diff
@@ -0,0 +1,292 @@
...
+      if (constraint_matrix.is_constrained(i))
+        v(i) += u(i);
+  };
+
+  return_op.vmult = [&constraint_matrix](Range &v, const Domain &u)
+  {
+    v = 0.;
+    for (auto i : v.locally_owned_elements())
+      if (constraint_matrix.is_constrained(i))
+        v(i) += u(i);
```
Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: you only touch every `v(i)` once, so no need for addition -- just set the element.

By the way, in all of these functions, wouldn't we want to assert that `v != u`?

File: include/deal.II/lac/constrained_linear_operator.h [hunk 28]
```diff
@@ -0,0 +1,292 @@
...
+  };
+
+  return_op.Tvmult = [&constraint_matrix](Domain &v, const Range &u)
+  {
+    v = 0.;
+    for (auto i : v.locally_owned_elements())
+      if (constraint_matrix.is_constrained(i))
+        v(i) += u(i);
```
Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: same here

File: include/deal.II/lac/constrained_linear_operator.h [hunk 29]
```diff
@@ -0,0 +1,292 @@
...
+ * @p exemplar, return an LinearOperator that is the projection to the
+ * subspace of constrained degrees of freedom.
+ *
+ * @author Mauro Bardelloni, Matthias Maier, 2015
+ *
+ * @relates LinearOperator
+ * @ingroup constraints
+ */
+template <typename Range, typename Domain>
+LinearOperator<Range, Domain> project_to_constrained_linear_operator(
```
Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: I think the name is misleading. If I understand the code right, what you really do is project onto the subspace of constrained degrees of freedom. You're not projecting one operator onto the subspace of "constrained linear operators". Maybe `project_to_constrained_subspace`?

Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: I realize that you append `_linear_operator` because you want to indicate that you return a linear operator. In essence, you're trying to scope the names of these functions. I think that would be better achieves by putting all of these linear operators into a namespace of their own.

File: include/deal.II/lac/constrained_linear_operator.h [hunk 30]
```diff
@@ -0,0 +1,292 @@
...
+    constraint_matrix.distribute(*k);
+
+    v += Ct * (right_hand_side - linop **k);
+
+    vector_memory.free(k);
+  };
+
+  // lambda capture expressions are a C++14 feature...
+  const auto apply_add = return_comp.apply_add;
+  return_comp.apply = [apply_add](Range &v)
```
Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: are you saying that C++11 is not enough? I don't think I understand the comment.

File: tests/lac/constraint_linear_operator_01.with_cxx11=on.output [hunk 31]
```diff
@@ -0,0 +1,10 @@
+
+DEAL::Solver stopped within 10 - 60 iterations
```
Comment (bangerth) [inline tests/lac/constraint_linear_operator_01.with_cxx11=on.output]: That's pretty generic. Can we either narrow the range?

Comment (tamiko) [inline tests/lac/constraint_linear_operator_01.with_cxx11=on.output]: No, and it is unimportant how many iterations the linear solver needs.

File: include/deal.II/lac/constrained_linear_operator.h (hunk 30 above)
Comment (tamiko) [inline include/deal.II/lac/constrained_linear_operator.h]: The following lambda capture expression is a C++14 feature:
```
return_comp.apply = [apply_add = return_comp.apply_add](Range &v)
...
```

File: include/deal.II/lac/constrained_linear_operator.h (hunk 29 above)
Comment (tamiko) [inline include/deal.II/lac/constrained_linear_operator.h]: And what should be a sensible namespace? ```LinearOperator``` is not possible to use any more. Furthermore, with what justification do we put those 4 functions into a namespace but leave the other dozen in the global namespace?

**Comment by tamiko** [comment]
*mhm*

File: include/deal.II/lac/constrained_linear_operator.h (hunk 29 above)
Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: I don't know -- it's just a suggestion to make the names of functions easier to understand.

We've often used plural names for namespaces in the past (`GridTools`, `VectorTools`, etc.)

File: tests/lac/constraint_linear_operator_01.with_cxx11=on.output (hunk 31 above)
Comment (bangerth) [inline tests/lac/constraint_linear_operator_01.with_cxx11=on.output]: I might disagree with this, but I'm not going to argue it.

**Comment by tamiko** [comment]
We'll try again in December.

File: include/deal.II/lac/constrained_linear_operator.h (hunk 30 above)
Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: Oh, I see. I'd say just remove the comment then, it's only understandable if you know where the author originally wanted to go with the code.

**Comment by bangerth** [comment]
You give up too quickly. I didn't argue the general usefulness of the patch, I'm simply talking about names and documentation. I know that's sometimes frustrating to deal with, but it surely is less work than you've already put into it, and you're a good enough software architect to know that these are exactly the issue that bite you in the long run if you don't address them in the beginning.
